docs/contributing/999-rfcs/2022-12-08-wing-inputs.md

```diff
@@ -0,0 +1,202 @@
+
+```toml
+# wing.toml
+bucketCount = 3
+```
+
+```js
+// main.w
+bring cloud;
+let bucketCount = cloud.Input("bucketCount");
+
+let buckets = MutArray<cloud.Bucket>[];
+
+for i in 0..buckets.asNum() {
+  buckets.push(new cloud.Bucket() as "bucket_${i}");
+}
+```
+
+
+Additionally lets imagine we wanted to create resources based on feature flags provided at compile time.
+
+```toml
+email-list = "one@email.com, two@gmail.com"
+
+[ features ]
+notifications = false
+```
+
+```js
+let enableNotifications = cloud.Input("/features/notifications");
+let emailList = cloud.Input("/email-list"); // comma separated string
+
+let inbox = new cloud.Bucket() as "inbox";
+
+// conditionally provision infrastructure based on enableNotification
+if enableNotifications.asBool() {
+	inbox.onCreate(inflight () => {
+		// use emailList input inflight
+		for email in emailList.asStr().split(",") {
+			// ...
+		}
+	});
+}
+```
+
+### Inputs via CLI Flags
+
+As an alternative to creating a `wing.toml` file, non-secure inputs can be defined via the cli.
+
+```bash
+wing compile main.w --input /feature/notifications=true --input /email-list=stuff
+```
+
+### Platform specific inputs
+
+In some cases input values need to vary based on the target Platform. We can imagine we may have a input required for a Platform names: "TerraformS3Backend" which requires 2 inputs at minimum, `tf-s3-backend/bucket` and `tf-s3-backend/state-file-name` the non app providing Platforms are able to add inputs during the `preSynth` hook.
+
+```js
+export class TerraformS3Backend implements IPlatform {
+  preSynth(app: App): void {
+    const rootKey = "/tf-s3-backend";
+	app.addInput("S3BackendBucket", {name: `${rootKey}/bucket`});
+	app.addInput("S3BackendStateFile", {name: `${rootKey}/state-file`});
+  }
+}
+```
+
+Compiling the following platform with missing inputs will result in an error message like such:
+```
+Missing required inputs:
+/tf-s3-backend/bucket
+/tf-s3-backend/state-file-name
+```
+
+To satisfy the required inputs we would need to create a `wing.toml` with those required values
+```toml
+[ tf-s3-backend ]
+bucket = "my-wing-bucket"
+state-file = "my state file"
+```
+
+### Creating inputs in preflight
+
+Sometimes we may just want to provide an input defined within our Wing application. Take for instance the following 
+```js
+let testEmail = new cloud.Input(name: "/test/email", defaultValue: "test@fake.com");
+
+new cloud.Function(inflight() => {
+	// some code that uses the testEmail
+});
+```
+
+While more adhoc than creating an `wing.toml` file, this should be fine in most non-secure cases.
+### Secure Inputs
+
+Some inputs contain sensitive data and thus should be treated with enhanced scrutiny. I.E. Inputs with sensitive data should not be created by preflight code in order to keep any trace of them from state files. 
```

Comment:
Not sure I follow this logic. The fact that I defined the input resource in preflight doesn't imply that it will store the value in state files.

It just means that we want the value to be read in inflight no?

Reply:
Wouldn't it either end up in state or in synthesized files? Neither is great for secure inputs. So I'm wondering if this should be explicitly discouraged?